### PR TITLE
planner: fix update privilege error

### PR DESCRIPTION
**Pull request by haplone**
### What problem does this PR solve? 
[issue10028](https://github.com/pingcap/tidb/issues/10028)
```sql
update ap.record t inner join tp.record tt on t.id=tt.id  set t.name=tt.name;
```

```
ERROR 1105 (HY000): privilege check fail
```

### What is changed and how it works?

get the right database name for update privilege check in planner/core/logical_plan_builder.go#buildUpdate

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

we need to fix in release-2.0 release-2.1 by another pr(buildUpdate logic has changed)

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2611,6 +2614,13 @@ func (b *PlanBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.A
 		col := assign.Col
 
 		dbName := col.DBName.L
+		// With sql below we need to get database by table alia name
```
Comment (xiekeyi98) [inline planner/core/logical_plan_builder.go]: What the `alia` stand for?
```suggestion
		// With sql below, we need to get database by the table alias name.
```

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -2611,6 +2614,13 @@ func (b *PlanBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.A
...
+		// update ap.record t inner join tp.record tt on t.id=tt.id  set t.name=tt.name
```
Comment (xiekeyi98) [inline planner/core/logical_plan_builder.go]:
```suggestion
		// issue#10028
```

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -2611,6 +2614,12 @@ func (b *PlanBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.A
 		col := assign.Col
 
 		dbName := col.DBName.L
+		// To solve issue#10028, we need to get database name by the table alias name.
+		for _, tbl := range tableList {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: will it be faster if we pre-build a `map[tblName]DBName` and save this `for` loop?

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -2611,6 +2614,12 @@ func (b *PlanBuilder) buildUpdateLists(tableList []*ast.TableName, list []*ast.A
...
+			if strings.EqualFold(tbl.Name.L, col.TblName.L) {
+				dbName = tbl.DBInfo.Name.L
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: BTW, after finding a dbName, we should `break`.

**Comment by zz-jason** [comment]
@haplone Please cherry pick this commit to the release-2.1 branch

**Comment by haplone** [comment]
@zz-jason 
I found branch release 2.1 on tidb has an error,
`the mysql.UpdatePriv should be mysql.SelectPriv`
 if it is fixed ,we could solve this issue in release 2.1 by cherry pick , or I have to make a new pr:
https://github.com/pingcap/tidb/blob/release-2.1/planner/core/logical_plan_builder.go#L2153
```go
func (b *planBuilder) buildUpdate(update *ast.UpdateStmt) (Plan, error) {
       // ...
        for _, t := range tableList {
		dbName := t.Schema.L
		if dbName == "" {
			dbName = b.ctx.GetSessionVars().CurrentDB
		}
		// the mysql.UpdatePriv should be mysql.SelectPriv
		b.visitInfo = appendVisitInfo(b.visitInfo, mysql.UpdatePriv, dbName, t.Name.L, "")
	}
```

**Comment by zz-jason** [comment]
seems we need to cherry pick https://github.com/pingcap/tidb/pull/8376 to release 2.1 first @tiancaiamao

**Pull request merged**